docs: Add initial intro to meta map (#1085)
nf-core/nf-co.re: markdown/developers/adding_modules.md

nextflow run tests/modules/<nameofmodule> -entry test_<nameofmodule> -c tests/config/nextflow.config -stub-run
```

> ## Into the `meta` map

Reviewer: ```suggestion
## What is the `meta` map?
```

> As you've noticed most modules take an input of

Reviewer 2: ```suggestion
In nf-core DSL2 pipelines, to add sample information and metadata, we use a meta variable. This avoids the need to create separate channels for each new characteristic. 
The meta variable can be passed down to processes as a tuple of the channel containing the actual samples, e.g. FastQ files, and the meta variable. 
The `meta map` is a [groovy map](https://www.tutorialspoint.com/groovy/groovy_maps.htm), which is like a python dictionary.
```

> 

Reviewer 2: ````suggestion
```nextflow
[id: 'test', single_end: false]
```
Thus, the information can be accessed by the key i.e. `meta.id`

The meta variable can be passed down to processes as a tuple of the channel containing the actual samples, e.g. FastQ files, and the meta variable.
````

```nextflow
input:
tuple val(meta), path(reads)
```

> You have probably also noticed, this pattern doesn't work out of the box with [fromFilePairs](https://www.nextflow.io/docs/edge/channel.html#fromfilepairs)

Reviewer 2: ```suggestion
This pattern doesn't work out of the box with [fromFilePairs](https://www.nextflow.io/docs/edge/channel.html#fromfilepairs)
```

The difference between the two:

```nextflow
// fromFilePairs
filepairs = [SRR493366, [/my/data/SRR493366_1.fastq, /my/data/SRR493366_2.fastq]]

// meta map
meta_map = [[id: 'test', single_end: false], // meta map
            [/my/data/SRR493366_1.fastq, /my/data/SRR493366_2.fastq]]
```

As you can see the difference, they are both [groovy lists](https://www.tutorialspoint.com/groovy/groovy_lists.htm).

> However, the filepairs just has a `val` that is a string, where as the `meta_map` the first value in the list, is a [groovy map](https://www.tutorialspoint.com/groovy/groovy_maps.htm), which is like a python dictionary.

Reviewer 2: ```suggestion
However, the filepairs just has a `val` that is a string, where as the `meta_map` the first value in the list, is a [groovy map](https://www.tutorialspoint.com/groovy/groovy_maps.htm)
```

Author: @asthara10 What's the issue with the reference to python dictionaries?

Author: Lovely, I missed that!

> The only required value is `id` for most of the modules, however they occasionally need things like `single_end`.

Reviewer 2: ```suggestion
The only required value is `meta.id` for most of the modules, however, they ususally contain fields like `meta.single_end` and `meta.strandedness`
```

> ### Common patterns

Reviewer 2: ```suggestion
### Generating the `meta map`
```

Author: Going to leave this, because I'm trying to collect some common nf-core nextflow patterns here. I don't think this is their final place, but it's better to collect them.

> The `meta map` is generated from [create_fastq_channel function in the input_check subworkflow](https://github.com/nf-core/rnaseq/blob/587c61b441c5e00bd3201317d48b95a82afe6aaa/subworkflows/local/input_check.nf#L23-L45) of most nf-core pipelines.

Reviewer 2: ```suggestion
The `meta map` is generated with [create_fastq_channel function in the input_check subworkflow](https://github.com/nf-core/rnaseq/blob/587c61b441c5e00bd3201317d48b95a82afe6aaa/subworkflows/local/input_check.nf#L23-L45) of most nf-core pipelines. Where the meta information is easily extracted from a samplesheet that contains the input file paths.
```

nextflow run tests/modules/<nameofmodule> -entry test_<nameofmodule> -c tests/config/nextflow.config -stub-run
```

## What is the `meta` map?

> In nf-core DSL2 pipelines, to add sample information and metadata, we use a meta variable. This avoids the need to create separate channels for each new characteristic. 

Reviewer 3: ```suggestion
In nf-core DSL2 pipelines, to add sample-specific information and metadata that is carried throughout the pipeline, we use a meta variable. This avoids the need to create separate channels for each new characteristic. 
```

The meta variable can be passed down to processes as a tuple of the channel containing the actual samples, e.g. FastQ files, and the meta variable. 

> The `meta map` is a [groovy map](https://www.tutorialspoint.com/groovy/groovy_maps.htm), which is like a python dictionary.

Reviewer 4: Should we show the whole channel structure at the beginning before introducing only the map e.g. `[[id: 'test', single_end: false], file]` explaining that the first item is the map and the second the file(s) itself or would it be more confusing?

Reviewer 4: ```suggestion
The `meta map` is a [groovy map](https://www.tutorialspoint.com/groovy/groovy_maps.htm), which is like a python dictionary, as shown below:
```

Author: > Should we show the whole channel structure at the beginning before introducing only the map e.g. `[[id: 'test', single_end: false], file]` explaining that the first item is the map and the second the file(s) itself, or would it be more confusing?

Hmm, it could work, could you write out what you're thinking? I don't know what the difference between the two would be.

> ```nextflow

Reviewer 5: ````suggestion

```nextflow
````

> ```nextflow
> [id: 'test', single_end: false]
> ```

Reviewer 5: Pretty sure markdown linting needs newlines around code blocks? It used to anyway 😅 

````suggestion

```nextflow
[id: 'test', single_end: false]
```

````

> Thus, the information can be accessed by the key i.e. `meta.id`

Reviewer 3: ```suggestion
Thus, the information can be accessed within processes and `module.conf` files with the key i.e. `meta.id`
```

The meta variable can be passed down to processes as a tuple of the channel containing the actual samples, e.g. FastQ files, and the meta variable.

> ```nextflow

Reviewer 5: ````suggestion

```nextflow
````

// meta map
meta_map = [[id: 'test', single_end: false], // meta map
            [/my/data/SRR493366_1.fastq, /my/data/SRR493366_2.fastq]]
```

As you can see the difference, they are both [groovy lists](https://www.tutorialspoint.com/groovy/groovy_lists.htm).
However, the filepairs just has a `val` that is a string, where as the `meta_map` the first value in the list, is a [groovy map](https://www.tutorialspoint.com/groovy/groovy_maps.htm), which is like a python dictionary.
The only required value is `meta.id` for most of the modules, however, they usually contain fields like `meta.single_end` and `meta.strandedness`

### Common patterns

The `meta map` is generated with [create_fastq_channel function in the input_check subworkflow](https://github.com/nf-core/rnaseq/blob/587c61b441c5e00bd3201317d48b95a82afe6aaa/subworkflows/local/input_check.nf#L23-L45) of most nf-core pipelines. Where the meta information is easily extracted from a samplesheet that contains the input file paths.

#### Generating a `meta map` from file pairs

Sometimes you want to use nf-core modules in small scripts. You don't want to make a samplesheet, or maintain a bunch of validation.

> For example here's an example script to run fastqc

Reviewer 4: ```suggestion
For instance, here's an example script to run fastqc
```
